src/azure-cli-testsdk/azure/cli/testsdk/scenario_tests/recording_processors.py

```diff
@@ -203,3 +203,25 @@ def process_request(self, request):
         import re
         request.uri = re.sub('(?<!:)//', '/', request.uri)
         return request
+
+
+class ContentLengthProcessor(RecordingProcessor):
+    # After replacing body content with processors like GeneralNameReplacer and SingleValueReplacer,
+    # Content-Length header will not match the actual length of body.
+    # We should re-compute Content-Length header with the new body.
+
+    def process_request(self, request):
+        # azure-core currently doesn't check request's Content-Length, but we fix it anyway.
+        if is_text_payload(request) and request.body and 'Content-Length' in request.headers:
+            body = str(request.body, 'utf-8') if isinstance(request.body, bytes) else str(request.body)
+            request.headers['Content-Length'] = str(len(body))
+        return request
+
+    def process_response(self, response):
+        # If Content-Length doesn't match the body,
+        # azure.core.pipeline.transport._requests_basic.StreamDownloadGenerator
+        # fails with azure.core.exceptions.IncompleteReadError:
+        # https://github.com/Azure/azure-sdk-for-python/pull/20888
+        if is_text_payload(response) and response['body']['string'] and 'content-length' in response['headers']:
+            response['headers']['content-length'][0] = str(len(response['body']['string']))
```

Comment:
Very good question. Because `request` is a `vcr.request.Request` object while `response` is a `dict`. `vcrpy` defines different data structures for them, though I don't know why.